File: em-april-2026/dsa/leetcode_practice_problems/unsorted/01419_min_num_of_frogs_croaking/main.py
class Solution:
    def minNumberOfFrogs(self, croakOfFrogs) -> int:
        c = 0
        r = 0
        o = 0
        a = 0
        k = 0
        current_frogs = 0
        max_frogs = 0

        for ch in croakOfFrogs:
            # 1. Increment the specific letter we just saw
            if ch == "c":
                c = c + 1
                current_frogs = current_frogs + 1
                max_frogs = max(max_frogs, current_frogs)
            elif ch == "r":
                r = r + 1
            elif ch == "o":
                o = o + 1
            elif ch == "a":
                a = a + 1
            elif ch == "k":
                k = k + 1
                current_frogs = current_frogs - 1
            else:
                return -1  # Invalid character

            # 2. The Golden Rule of Croaking:
            # A letter can never be counted more times than the letter before
            # it.
            if (c < r or r < o or o < a or a < k):
                return -1
        # 3. Ensure no frogs are left mid-croak
        if current_frogs == 0:
            return max_frogs
        return -1


def main():
    s = Solution()
    input = "crcoakroak"
    # input = "croakcroak"
    output = s.minNumberOfFrogs(input)
    print(f"input = {input}, output = {output}")


if __name__ == "__main__":
    main()

File: em-april-2026/dsa/leetcode_practice_problems/unsorted/01419_min_num_of_frogs_croaking/main.cpp
#include <algorithm>
#include <cstring>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int minNumberOfFrogs(string croakOfFrogs) {
        int c = 0, r = 0, o = 0, a = 0, k = 0;
        int current_frogs = 0, max_frogs = 0;

        for (char ch : croakOfFrogs) {
            // 1. Increment the specific letter we just saw
            if (ch == 'c') {
                c++;
                current_frogs++;
                max_frogs = max(max_frogs, current_frogs);
            } else if (ch == 'r')
                r++;
            else if (ch == 'o')
                o++;
            else if (ch == 'a')
                a++;
            else if (ch == 'k') {
                k++;
                current_frogs--;
            } else {
                return -1;  // Invalid character
            }

            // 2. The Golden Rule of Croaking:
            // A letter can never be counted more times than the letter before
            // it.
            if (c < r || r < o || o < a || a < k) {
                return -1;
            }
        }

        // 3. Ensure no frogs are left mid-croak
        if (current_frogs == 0) {
            return max_frogs;
        }

        return -1;
    }
};

int main(int argc, char* argv[]) {
    Solution s;
    auto input = "crcoakroak";
    if (argc > 1) {
        if (strlen(argv[1]) < 1) {
            cerr << "bad input: " << argv[1] << endl;
            return 1;
        } else {
            input = argv[1];
        }
    }
    // auto input = "croakcroak";
    auto output = s.minNumberOfFrogs(input);
    cout << "input = " << input << ", output = " << output << endl;
    return 0;
}
